File: ios/Mailrs/Wire/RecipientAutocomplete.swift
import Foundation

/// The token arithmetic under the To field's contact suggestions.
///
/// A To line is comma/semicolon-separated; suggestions apply to the
/// token the cursor is in — in practice the last one, since address
/// entry is append-shaped. Completion replaces that token with the
/// picked contact's addr-spec and reopens the list for the next
/// entry with a trailing separator.
enum RecipientAutocomplete {
    /// What the person is currently typing: the text after the last
    /// separator, trimmed. Empty when they just finished an entry.
    static func currentToken(of text: String) -> String {
        guard let last = text.lastIndex(where: { $0 == "," || $0 == ";" }) else {
            return text.trimmingCharacters(in: .whitespaces)
        }
        return String(text[text.index(after: last)...])
            .trimmingCharacters(in: .whitespaces)
    }

    /// Whether a suggestion query is worth a request: two characters,
    /// same floor as search, and not already a complete address —
    /// suggesting for "alice@example.com" is answering a question that
    /// was already answered.
    static func shouldSuggest(for token: String) -> Bool {
        token.count >= 2 && !token.contains("@")
    }

    /// The text with the in-progress token replaced by the picked
    /// contact's bare address, ready for the next entry.
    static func completing(_ text: String, with contact: String) -> String {
        let email = SenderName.extractEmail(contact)
        guard let last = text.lastIndex(where: { $0 == "," || $0 == ";" }) else {
            return email + ", "
        }
        let kept = String(text[...last])
        return kept + " " + email + ", "
    }
}

File: android/app/src/main/kotlin/jp/golia/mailrs/wire/RecipientAutocomplete.kt
package jp.golia.mailrs.wire

/**
 * The token arithmetic under a recipient field's suggestions.
 *
 * Ported from `ios/Mailrs/Wire/RecipientAutocomplete.swift`. A
 * recipient line is comma/semicolon-separated; suggestions apply to the
 * token the cursor is in — in practice the last one, since address entry
 * is append-shaped. Completing replaces that token with the picked
 * contact's bare address and leaves a separator ready for the next.
 */
object RecipientAutocomplete {

    /**
     * What is being typed: the text after the last separator, trimmed.
     * Empty when the person has just finished an entry.
     */
    fun currentToken(text: String): String {
        val last = text.indexOfLast { it == ',' || it == ';' }
        if (last < 0) return text.trim()
        return text.substring(last + 1).trim()
    }

    /**
     * Whether a query is worth a request: two characters, the same floor
     * as search, and not already a complete address — suggesting for
     * "alice@example.com" answers a question that was already answered.
     */
    fun shouldSuggest(token: String): Boolean = token.length >= 2 && !token.contains("@")

    /** The line with the in-progress token replaced, ready for the next. */
    fun completing(text: String, contact: String): String {
        val email = SenderIdentity.emailOf(contact)
        val last = text.indexOfLast { it == ',' || it == ';' }
        if (last < 0) return "$email, "
        return text.substring(0, last + 1) + " " + email + ", "
    }
}
